app/controllers/graphql_controller.rb

```diff
@@ -0,0 +1,111 @@
+# frozen_string_literal: true
+
+class GraphqlController < ApplicationController
+  # Note 22/02/10 - Nimmo
+  # This controller is short but important.
+  # Parses an incoming request header to build a Context hash (authentication)
+  # and forwards that to the graphql Query, w/ Variables and Operation Name.
+  # Possible operation names: Query, Mutation, Union etc
+
+  disable_filters
+
+  def execute
+    variables = prepare_variables(params[:variables])
+    query = params[:query]
+    operation_name = params[:operationName]
+    context = {
+      current_user: current_user,
+      in_admin_mode?: in_admin_mode?
+    }
+    result = MushroomObserverSchema.execute(
+      query,
+      variables: variables,
+      context: context,
+      operation_name: operation_name
+    )
+
+    render(json: result)
+  rescue StandardError => e
+    raise(e) unless Rails.env.development?
+
+    handle_error_in_development(e)
+  end
+
+  private
+
+  # https://www.howtographql.com/graphql-ruby/4-authentication/
+  # Decrypt the current user from token stored in the header (not the session)
+  # Note: There may not be a token!
+  def current_user
+    return unless token.is_a?(Hash)
+    raise("Token missing user_id") unless token.key?("user_id")
+
+    user = User.safe_find(token["user_id"])
+    raise("User not found") unless user
+    # Should we raise an error here? Silently giving no current_user:
+    return unless user.verified
+
+    user
+  end
+
+  def in_admin_mode?
+    return false unless token.is_a?(Hash)
+
+    token["in_admin_mode"]
+  end
+
+  def token
+    ::Token.decrypt_from_header(http_auth_header).hash
```

Comment:
Definitely memoize this one, too.

Reply:
Just attempted this - 
It works if I add `@token ||= ::Token.decrypt_from_header(http_auth_header).hash` but not if i'm replacing `token` with `@token` everywhere in the other methods. I don't understand how it's helping if i don't use the instance var everywhere?